postgresql/database.go

```diff
@@ -622,3 +633,37 @@ func (d *database) getPrimaryKey(tableName string) ([]string, error) {
 
 	return tableSchema.PrimaryKey, nil
 }
+
+// waitForConnection tries to execute the connectFn function, if connectFn
+// returns an error, then waitForConnection will keep trying until connectFn
+// returns nil. Maximum waiting time is 5s after having acquired the lock.
+func waitForConnection(connectFn func() error) error {
+	// This lock ensures first-come, first-served and prevents opening too many
+	// file descriptors.
+	waitForConnMu.Lock()
+	defer waitForConnMu.Unlock()
+
+	// Minimum waiting time.
+	waitTime := time.Millisecond * 10
+
+	// Waitig 5 seconds for a successful connection.
+	for timeStart := time.Now(); time.Now().Sub(timeStart) < time.Second*5; {
+		if err := connectFn(); err != nil {
```

Comment:
The first iteration will try `connectFn()` with to delay.
The second iteration will try again after `waitTime` (10ms).
Third iteration will double `waitTime` (20ms). `waitTime` will keep on increasing while its value is still less than `500ms`.